lua: add withPackages function

teto opened the pull request:
First step towards more automation similar to the haskell backend.
Extracted from https://github.com/NixOS/nixpkgs/pull/33903 to make it easier to review.

###### Motivation for this change

I want to be able to create lua environments with lua.withPackages().

If some things appear awkward, more context at https://github.com/NixOS/nixpkgs/pull/33903 can help or just ask and I will try to answer if I can. I started this 1 year and half ago with no nix knowledge so some things can be improved (I would like to have a lua `lib` component as haskell does, my shell hooks might not be the best) but it kinda works.

Something I did not dare do is merge the three lua 5.1/ lua 5.2 /lua5.3 interpreters into one which would remove some duplication. If I have the greenlight I want to do it.

You can test with sthg like:
`NIX_PATH="nixpkgs=/home/teto/nixpkgs2:$NIX_PATH" nix-shell -p 'lua5_3.withPackages(ps: [ ps.lpeg ps.cqueues ])'`

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

teto committed:
lua: add withPackages function

First step towards more automation similar to the haskell backend.
Follow up of https://github.com/NixOS/nixpkgs/pull/33903

7c6f434c commented:
Please unbreak LuaJIT evaluation back…

teto commented:
So I am running locally the command from ofborg's readme with -vvvvv in order to diagnose the issue  

```
$ curl -o outpaths.nix https://raw.githubusercontent.com/NixOS/ofborg/released/ofborg/src/outpaths.nix
$ GC_INITIAL_HEAP_SIZE=4g nix-env -f ./outpaths.nix -qaP --no-name --out-path --arg checkMeta true -vvvvv > out-paths

```
but it fails with 
```
instantiated 'source' -> '/nix/store/25pdni30v75m9b85dygai4yvw4ydc3h4-source.drv'
processing attribute 'stdenv'
processing attribute 'strictDeps'
processing attribute 'system'
processing attribute 'userHook'
instantiated '9pfs-20150918' -> '/nix/store/4h2sk33qpk3hjap8mkjr0j1nixg0ynr6-9pfs-20150918.drv'
evaluating attribute '__splicedPackages'
error: out of memory
```

@grahamc any idea on how to properly debug what part of the code is responsible for the error (like tracing the name of the function etc.)

teto commented:
ok found the problem @7c6f434c could you have a look at the overall PR please?  The shell script might need some changes/cleanup. I think I should rebase it on staging (what is staging-next ?).

7c6f434c commented on pkgs/development/interpreters/lua-5/setup-hook.sh:
```diff
@@ -0,0 +1,72 @@
+    cd -
+}
+
+addEnvHooks "$hostOffset" startLuaEnvHook
```
Won't this add paths without checking in `luaPathsSeen`?

teto replied:
luaPathsSeen is effective when using the wrapper. The setup-hook is a bit different. With a naive version adding every possible folder for any package, it could end up too long. The version thus checks if the folder exists and only in that case it will add it to LUA_PATH. The "requiredLuaModules" calls `unique` function so it should prevent duplicates from appearing.

7c6f434c replied:
Are these called for  `requiredLuaModules` or for `buildInputs`, though?

(Also, is there a reasonable way to just use wrappers during normal builds, to reduce code duplication, as well as confusion because of inexact duplication?)

(also, if `luaPathsSeen` is not relevant for this file, why the comment mentions it?)

7c6f434c replied:
Do you know if any duplication happens in case of multiple buildInputs propagate the same path?

teto replied:
I've added a check against duplicates but this hook is run when building user environments so IIRC it is run only on the environment folder.

7c6f434c commented:
Re: target branch — I think if building `lua*Packages` and `torch*Packages` takes reasonable time locally, it can go to `master`; otherwise yes, `staging` is better.

teto commented:
lua and torch packages are usually fast to compile so master should be fine IMO. Rest assured that should any problem arise I will promptly fix them as I would like to follow up with other PRs as I described in the first updated post of https://github.com/NixOS/nixpkgs/pull/33903

teto commented:
I removed the luaWrap for now as it doesn't look necessary for now, and what I want to do first is to update the packages so that they use lua environments. It will make my other changes easier to rebase/submit.
Could you have another look please ?

teto committed:
lua: simplify setup hook in wrapper.nix

check for duplicate as well.

7c6f434c commented:
Nice, GitHub shows me the list of changes including a file that has empty diff. What the… convenient and consistent UI.

I like the current version much better; I am not sure if handling repeated buildInputs in the shell code is worth the effort (maybe we could just see how it goes).

Do I understand correctly that you expect this version of interpreter packages to stick for a while?

I guess if this part is stable, I could reread it once more and merge it…

teto commented:
> Do I understand correctly that you expect this version of interpreter packages to stick for a while?

Apart from merging the lua5.1/5.2/5.3 interpreters into one - which is low priority for me - my next steps would be:
1/ to make some packages  (neovim/mpv/awesome mostly) depend on withPackages.
2/ Introduce buildLuarocksPackage and the "generated" packages ecosystem. I had first intended to completely replace the current infrastructure but in  https://github.com/NixOS/nixpkgs/pull/33903, I was advised to deploy my changes under a different name to ease upstreaming.

In terms of building, I will need to reintroduce a luaWrap (needed by some programs) but having the withPackages already available in nixos-unstable makes developement easier (and upgrade of some packages possible) so I am looking forward to this being merged.

7c6f434c commented on pkgs/development/interpreters/lua-5/setup-hook.sh:
```diff
@@ -0,0 +1,50 @@
+    # export only if the folder exists else LUA_PATH grows too big
+    if  [ ! -d "$topDir" ]; then return; fi
+    # check for duplicate
+    if [[ $LUA_PATH = *"$pattern"* ]]; then return; fi
```
This check requries that patterns cannot be substrings of each other (which is true; not sure if worth a comment); but shouldn't either the variable to check depend on `varName` or `;$LUA_PATH` just be checked  against `;$topDir/` ?

teto replied:
right it should be topDir/$pattern, let me fix it, At the same time, I never had any trouble with duplicate buildInputs so I am tempted to just remove the line.

teto replied:
ha no sorry pattern already contains the basefolder so it should be ok as is. I got confused because pattern don't have the same meaning in addToLuaSearchPathWithCustomDelimiter (absolute path) and addToLuaPath (relative path).

7c6f434c replied:
What I meant is that if `;$LUA_PATH` contains `;$topDir` we know that the bsae directory has been processed already.

teto replied:
ok I've dropped the check and renamed pattern to abs(olute)Pattern to make it easier it has a different meaning. No need to troll myself.

teto replied:
checking just against topdir would not be enough since several patterns are checked against the same topdir. The first matching pattern would discard others while they may be valid as well.

teto committed:
lua setup hook: dropping the check for duplicate

also makes clearer that in one case pattern is a full path while in the other it
is relative.

7c6f434c commented:
@GrahamcOfBorg build love_11 luaPackages.luarocks luaPackages.luarocks-nix luaPackages.http

7c6f434c commented:
GitHub: «10 checks passed» — that's counting only eval checks after I intentionally waited for a build to pass. Sigh.

vcunat commented:
Cross-ref: https://github.com/NixOS/nixpkgs/pull/55345
